thunder/core/transform_common.py

```diff
@@ -463,133 +463,289 @@ def is_functionalizable(bsym: BoundSymbol) -> bool:
+
+
 def collect_required_copy_bsyms_for_args(
     trace: Trace,
-    removed_copy_bsyms: list[BoundSymbol],
-) -> tuple[list[BoundSymbol], dict[VariableInterface, TensorProxy]]:
-    from thunder.core.utils import consumers as get_consumers
-    from thunder.core.utils import find_producer_symbols
-    from thunder.torch import _syms_returning_views
+    removed_copy_bsyms: list[tuple[int, BoundSymbol]],
+) -> list[BoundSymbol]:
 
-    flat_args = tuple(
-        filter(
-            lambda p: isinstance(p, TensorProxy),
-            tree_flatten((trace.args, trace.kwargs))[0],
-        )
+    flat_args: tuple[TensorProxy, ...] = tuple(
+        filter(lambda p: isinstance(p, TensorProxy), tree_flatten((trace.args, trace.kwargs))[0])
     )
     trace_args_set = ProxyDict()
-    for a in flat_args:
-        trace_args_set[a] = a
-
-    copy_dst_to_copy_bsym: dict[VariableInterface, BoundSymbol] = {}
-    required_copy_bsyms: list[BoundSymbol] = []
-    unused_copy_bsyms: list[BoundSymbol] = []
-    inplace_args: list[TensorProxy] = []
-    copy_destinations: list[TensorProxy] = []
-    for idx, bsym in enumerate(removed_copy_bsyms):
-        if (copy_to := bsym.flat_proxy_args[1]) in trace_args_set:
-            required_copy_bsyms.append(bsym)
-            copy_dst_to_copy_bsym[variableify(copy_to)] = bsym
-            inplace_args.append(copy_to)
+    for i, a in enumerate(flat_args):
+        trace_args_set[a] = i
+
+    dst_to_copy_bsym: dict[VariableInterface, tuple[int, BoundSymbol]] = {}
+    unused_copy_bsyms: list[tuple[int, BoundSymbol]] = []
+    for idx, copy_bsym in removed_copy_bsyms:
+        if (copy_dst := copy_bsym.flat_proxy_args[1]) in trace_args_set:
+            copy_src = _get_first_tensor_arg(copy_bsym)
+            dst_to_copy_bsym[variableify(copy_dst)] = idx, copy_bsym
         else:
-            unused_copy_bsyms.append(bsym)
+            unused_copy_bsyms.append((idx, copy_bsym))
+
+    producer_map = producers(trace)
+    new_dst_to_copy_bsym: dict[VariableInterface, tuple[int, list[BoundSymbol]]] = {}
+    for idx, copy_bsym in unused_copy_bsyms:
+        copy_src = _get_first_tensor_arg(copy_bsym)
+        copy_dst: TensorProxy = copy_bsym.flat_proxy_args[1]
+        optional_prod_bsym = _get_prod_bsym_with_arg(copy_dst, producer_map, trace_args_set)
+        if optional_prod_bsym is None:
+            continue
 
-        copy_destinations.append(copy_to)
+        new_copy_dst = _get_first_tensor_arg(optional_prod_bsym)
 
-    consumers = get_consumers(trace)
-    view_args = []
-    for a in filter(lambda a: variableify(a) not in copy_dst_to_copy_bsym, flat_args):
-        first_consumer = consumers[a][0]
-        if first_consumer.sym in _syms_returning_views:
-            view_args.append(a)
-    stop_proxies = tuple(inplace_args + view_args)
+        new_bsyms: list[BoundSymbol] = []
+        reshaped_copy_src = copy_src
+        with tracectx(trace):
+            if copy_src.shape != new_copy_dst.shape:
```

Comment:
So here, is my understanding correct that we adapt the shape to cope with .view in between but we don't need broadcasting because we don't do that yet (and because our `copy_` is not `torch.Tensor.copy_`)?

Reply:
I think there're two reasons

- `torch.Tensor.copy_` is not covered in `thunder/torch/__init__.py`
- thunder.jit errors out if a view tensor of different numel from the original is the operand of an in-place op here: https://github.com/Lightning-AI/lightning-thunder/blob/e646e4fce78542410a77c8be7de38a659860b536/thunder/core/transform_common.py#L444-L451

Reply:
Thank you!